apple/internal/linking_support.bzl

```diff
@@ -51,6 +54,26 @@ def _debug_outputs_by_architecture(link_outputs):
         linkmaps = linkmaps,
     )
 
+def _new_executable_binary_provider(*, binary, cc_info, objc):
+    """Wrap the apple_common API of the same name to better support multiple Bazel versions.
+
+    Args:
+        binary: The `File` representing the binary.
+        cc_info: The `CcInfo` provider for the binary.
+        objc: The `apple_common.Objc` provider for the binary.
+    """
+    if _OBJC_PROVIDER_LINKING:
+        return apple_common.new_executable_binary_provider(
```

Comment:
it looks to me like these rules are the only consumers of this info, so theoretically we could switch to our own provider for this instead, if we did the only field we currently use is the `binary`